### DISCUSSION TITLE
New requirements for the input buffer

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#164] [Ideas]
This is a followup to the recent [discussion](https://github.com/ForthHub/discussion/discussions/161#discussioncomment-10406876).

The section [3.3.3.5 Input buffers](https://forth-standard.org/standard/usage#usage:inbuf) of Forth-2012 (as well as [Forth-2019 draft](http://www.forth200x.org/documents/forth19-1.pdf)) says:

> The address and length returned by [`SOURCE`](https://forth-standard.org/standard/core/SOURCE), the string returned by [`PARSE`](https://forth-standard.org/standard/core/PARSE), and directly computed input-buffer addresses are valid only until the text interpreter does I/O to refill the input buffer or the input source is changed.

A proposed change:
> The address and length returned by [`SOURCE`](https://forth-standard.org/standard/core/SOURCE), the string returned by [`PARSE`](https://forth-standard.org/standard/core/PARSE), and directly computed input-buffer addresses <del>are valid only</del> <ins>shall be valid</ins> until the text interpreter does I/O to refill the input buffer <ins>or restore the input source position (see [`RESTORE-INPUT`](https://forth-standard.org/standard/core/RESTORE-INPUT)),</ins> <del>or the input source is changed</del> <ins>or interpretation of the input source is complete</ins>.

It seems, reposition by means of `restore-input` should be mentioned anyway (it's an omission).

Consequences of this change:
  - a program is allowed to pass a character string that is a part of the input buffer via `include-file`, `included`, or `load` (directly or indirectly);
  - a program is allowed to use input-buffer addresses after a subsequent invocation of `include-file`, `included`, or `load`;
  - the method 1 mentioned in [A.11.6.1.1717 `INCLUDE-FILE`](https://forth-standard.org/standard/rationale#rat:file:INCLUDE-FILE) is not allowed for a standard Forth system anymore;
  - `buffer` and `block` are not allowed to assign the input buffer to another block until interpretation of this buffer is complete.

**Does anybody know of a standard Forth system (or perhaps a system guided by the standard) for which following that change is problematic?**

**Comment by SirWumpus** [comment]
Isn't there a separate proposal to deprecate `SAVE-INPUT` and `RESTORE-INPUT` as being problematic and rarely used? I recall we had a discussion about this some time ago.

**Reply by ruv** [reply]
Yes, that [proposal](https://forth-standard.org/proposals/obsolescence-for-save-input-and-restore-input) exists.  But until that proposal is accepted, `save-input` or `restore-input` shall be mentioned if a standard program must be aware of their effects.
